ksp_plugin/celestial_body.hpp

```diff
@@ -5,114 +5,65 @@
+    Instant const& current_time) const {
   CHECK(is_initialized());
-  return prolongation_;
+  return trajectory().EvaluateDegreesOfFreedom(current_time,
+                                               current_time_hint());
 }
 
-inline Trajectory<Barycentric> const& Celestial::prediction() const {
+inline Position<Barycentric> Celestial::current_position(
+    Instant const& current_time) const {
   CHECK(is_initialized());
-  return *CHECK_NOTNULL(prediction_);
+  return trajectory().EvaluatePosition(current_time, current_time_hint());
 }
 
-inline Trajectory<Barycentric>* Celestial::mutable_prediction() {
+inline Velocity<Barycentric> Celestial::current_velocity(
+    Instant const& current_time) const {
   CHECK(is_initialized());
-  return prediction_;
-}
-
-inline bool Celestial::has_prediction() const {
-  return prediction_ != nullptr;
-}
-
-inline void Celestial::CreateHistoryAndForkProlongation(
-    Instant const& time,
-    DegreesOfFreedom<Barycentric> const& degrees_of_freedom) {
-  history_ = std::make_unique<Trajectory<Barycentric>>(body_.get());
-  history_->Append(time, degrees_of_freedom);
-  prolongation_ = history_->NewFork(time);
-}
-
-inline void Celestial::ResetProlongation(Instant const& time) {
-  history_->DeleteFork(&prolongation_);
-  prolongation_ = history_->NewFork(time);
+  return trajectory().EvaluateVelocity(current_time, current_time_hint());
 }
 
-inline void Celestial::ForkPrediction() {
-  CHECK(prediction_ == nullptr);
-  prediction_ = mutable_prolongation()->NewFork(prolongation().last().time());
+inline MassiveBody const& Celestial::body() const {
+  return *body_;
 }
 
-inline void Celestial::DeletePrediction() {
-  prolongation_->DeleteFork(&prediction_);
+inline bool Celestial::has_parent() const {
+  return parent_ != nullptr;
 }
 
-inline void Celestial::WriteToMessage(
-    not_null<serialization::Celestial*> const message) const {
-  CHECK(is_initialized());
-  body_->WriteToMessage(message->mutable_body());
-  history_->WriteToMessage(
-      message->mutable_history_and_prolongation()->mutable_history());
-  prolongation_->WritePointerToMessage(
-      message->mutable_history_and_prolongation()->mutable_prolongation());
+inline Celestial const* Celestial::parent() const {
+  return parent_;
 }
 
-inline std::unique_ptr<Celestial> Celestial::ReadFromMessage(
-    serialization::Celestial const& message) {
-  auto celestial =
-      std::make_unique<Celestial>(MassiveBody::ReadFromMessage(message.body()));
-  celestial->history_ =
-      Trajectory<Barycentric>::ReadFromMessage(
-          message.history_and_prolongation().history(),
-          celestial->body_.get());
-  celestial->prolongation_ =
-      Trajectory<Barycentric>::ReadPointerFromMessage(
-          message.history_and_prolongation().prolongation(),
-          celestial->history_.get());
-  return celestial;
+inline void Celestial::set_parent(not_null<Celestial const*> const parent) {
```

Comment:
We could, but we still need a setter, unless UpdateCelestialHierarchy disappears. I'm not sure whether we need that one, but let's not look at this now.